Bluetooth: controller: Miscellaneous radio timing tuned

cvinayak opened the pull request:
- In the Tx to Rx switch, fix the delay which should be the
Rx ready delay plus the Tx chain delay and plus 4 us active
clock jitter compensation.
- We need to take RX chain delay and PHY used to receive
CONNECT_IND into account when calculating 1st connection
event offset.
- PA/LNA related timing calculations tuned for extended
advertising and scanning.
- Minor update to calls to get Tx chain delay in extended
advertising and scanning to use S8 encoding delays.

cvinayak committed:
Bluetooth: controller: Fix Tx-Rx switch delay calculation

In the Tx to Rx switch, fix the delay which should be the
Rx ready delay plus the Tx chain delay and plus 4 us active
clock jitter compensation.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

andrzej-kaczmarek commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -580,9 +580,17 @@ static inline void sw_switch(uint8_t dir_curr, uint8_t dir_next,
-			hal_radio_tx_chain_delay_ns_get(phy_curr, flags_curr));
+			hal_radio_rx_chain_delay_ns_get(phy_curr, flags_curr)) +
+			4;
+#endif
```
I do not really want to start another discussion about timing, but this particular change looks like a workaround.

tx_chain_delay should compensate for EVENT_END being triggered before end of PDU in the air (TX path delay from digital radio to antenna) which is 0.6us, that was correct. Once you changed to rx_chain_delay which is 9.4us, it reduces compensation too much and radio starts RX somewhere during preamble. It works on hw, because it can receive PDU with incomplete preamble. Apparently bsim cannot do this.

![image](https://user-images.githubusercontent.com/9692685/118472829-94903700-b709-11eb-915f-835ddadbdadf.png)

aescolar replied:
> Apparently bsim cannot do this.

In bsim, the phy model only considers a packet in air as possibly received if the receiver is open during the first bit of the preamble (this is a simplification, and probably a bit tougher than what some radios may be able to accept, which seems to include yours :) )

cvinayak replied:
What I am trying to understand here is, TXEN starts with first bit transmitted from transmitter and RADIO_END is the last bit received on the receiver. I am sure there is time spent between the first bit and last bit of the last byte of the PDU (CRC last byte). This (first bit to last bit of the last byte) is the amount of time that the receiver will have to subtract when turning around to transmitted back in order  to meet the TIFS which is end of last bit received to the transmit of first bit back. This is my definition of rx_chain_delay used when transmitting back, which depends on the transmitted PHY.

> Once you changed to rx_chain_delay which is 9.4us, it reduces compensation too much and radio starts RX somewhere during preamble.

In the code above, RXEN happens 16 us (10 us not compensated plus 10 us delayed minus the 4 us) delayed; note the ready_delay that is reduced by 10 us (for 1M PHY). This should, in theory, fail access address from being received unless radio is receiving earlier than the ready delay.

Now I am myself confused.

In thoery, then it should be

```
		delay = HAL_RADIO_NS2US_CEIL(
			hal_radio_rx_ready_delay_ns_get(phy_next, flags_next) +
			hal_radio_rx_chain_delay_ns_get(phy_curr, flags_curr)) +
			4;
```

Radio PPI should be triggered at (40 + 9.4  + 4) before the preamble. i.e.  (150 - 54), 96 us after the Tx Radio End. But for some reason, triggering the PPI at (40 - 9.4 + 4) before the preamble, (150 - 35), 115 us after the Tx Radio End still passes the conformance timing tests!

cvinayak committed:
Bluetooth: controller: PA/LNA radio timing tuned

PA/LNA related timing calculations tuned for extended
advertising and scanning.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Use S8 encoding Tx chain delay

Minor update to calls to get Tx chain delay in extended
advertising and scanning to use S8 encoding delays.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Fix 1st connection event offset for peripheral

We need to take RX chain delay and PHY used to receive
CONNECT_IND into account when calculating 1st connection
event offset.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
@andrzej-kaczmarek @kruithofa Changes to use correct chain delays, please review.

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -582,9 +582,13 @@ static inline void sw_switch(uint8_t dir_curr, uint8_t dir_next,
 	} else {
 		/* RX */
+
+		/* Calculate delay with respect to current and next PHY.
+		 */
```
```suggestion
		/* Calculate delay with respect to current and next PHY. */
```

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -582,9 +582,13 @@ static inline void sw_switch(uint8_t dir_curr, uint8_t dir_next,
-			hal_radio_tx_chain_delay_ns_get(phy_curr, flags_curr));
+			hal_radio_rx_ready_delay_ns_get(phy_next, flags_next) +
+			hal_radio_tx_chain_delay_ns_get(phy_curr, flags_curr)) +
+			4;
```
I see from the commit message that the `+ 4` is from jitter. Could we store this value in a named variable and use that? It improves readability to not have "magic numbers" in the code in my opinion.

cvinayak committed:
Bluetooth: Controller: Define a macro for active clock jitter value

Define a macro for active clock jitter value.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>
